cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -2055,6 +2035,116 @@ class AzureFileSystem::Impl {
+  /// \pre location.path does not end with a trailing slash.
+  Status DeleteFileOnContainer(const Blobs::BlobContainerClient& container_client,
+                               const AzureLocation& location, bool require_file_to_exist,
+                               const char* operation) {
+    DCHECK(!location.container.empty());
+    DCHECK(!location.path.empty());
+    DCHECK(!internal::HasTrailingSlash(location.path));
+    constexpr auto kFileBlobLeaseTime = std::chrono::seconds{15};
+
+    // If the parent directory of a file is not the container itself, there is a
+    // risk that deleting the file also deletes the *implied directory* -- the
+    // directory that is implied by the existence of the file path.
+    //
+    // In this case, we must ensure that the deletion is not semantically
+    // equivalent to also deleting the directory. This is done by ensuring the
+    // directory marker blob exists before the file is deleted.
+    std::optional<LeaseGuard> file_blob_lease_guard;
+    const auto parent = location.parent();
+    if (!parent.path.empty()) {
+      // We have to check the existence of the file before checking the
+      // existence of the parent directory marker, so we acquire a lease on the
+      // file first.
+      ARROW_ASSIGN_OR_RAISE(auto file_blob_lease_client,
+                            AcquireBlobLease(location, kFileBlobLeaseTime,
+                                             /*allow_missing=*/!require_file_to_exist));
+      if (file_blob_lease_client) {
+        file_blob_lease_guard.emplace(std::move(file_blob_lease_client),
+                                      kFileBlobLeaseTime);
+        // Ensure the empty directory marker blob exists before the file is deleted.
+        //
+        // There is not need to hold a lease on the directory marker because if
+        // a concurrent client deletes the directory marker right after we
+        // create it, the file deletion itself won't be the cause of the directory
+        // deletion. Additionally, the fact that a lease is held on the blob path
+        // semantically preserves the directory -- its existence is implied
+        // until the blob representing the file is deleted -- even if another
+        // client deletes the directory marker.
+        RETURN_NOT_OK(EnsureEmptyDirExists(container_client, parent, operation));
+      } else {
+        // File blob doesn't exist and require_file_to_exists is false, so just return OK.
+        // Trying to delete a file that doesn't exist, costs only one request to the
+        // server.
+        DCHECK(!require_file_to_exist);
+        return Status::OK();
+      }
+    }
+
+    auto blob_client = container_client.GetBlobClient(location.path);
+    Blobs::DeleteBlobOptions options;
+    if (file_blob_lease_guard) {
+      options.AccessConditions.LeaseId = file_blob_lease_guard->LeaseId();
+    }
+    try {
+      auto response = blob_client.Delete(options);
```

Comment:
Do we need to call `file_blob_lease_guard->BreakBeforeDeletion()` before `Delete()`?

Reply:
This was surprising to me as well. In the Azure Blobs API if I delete a blob that’s holding a lease I can release the lease on that path just fine. Re-creating a Blob on the same path immediately after deletion+release works just fine. This was a mess in the DataLake API, but is fine in the Blobs API.

Reply:
Wow. Could you add the information as a comment of here or `LeaseGuard::BreakBeforeDeletion()`?

Reply:
Alright. Adding a comment on `BreakBeforeDeletion`.